pennylane/templates/embeddings/iqp.py

```diff
@@ -0,0 +1,251 @@
+
+            @qml.qnode(dev)
+            def circuit(features=None, pattern=None):
+                IQPEmbedding(features=features, wires=range(3), pattern=pattern, n_repeats=3)
+                return [qml.expval(qml.PauliZ(w)) for w in range(3)]
+
+            res1 = circuit(features=[1., 2., 3.], pattern=pattern1)
+            res2 = circuit(features=[1., 2., 3.], pattern=pattern2)
+
+            assert np.allclose(res1, res2)
+
+        **Repeating the embedding**
+
+        The embedding can be repeated by specifying the ``n_repeats`` argument:
+
+        .. code-block:: python
+
+            @qml.qnode(dev)
+            def circuit(features=None):
+                IQPEmbedding(features=features, wires=range(3), n_repeats=4)
+                return [qml.expval(qml.PauliZ(w)) for w in range(3)]
+
+            circuit(features=[1., 2., 3.])
+
+        Every repetition uses exactly the same quantum circuit.
+
+    """
+    #############
+    # Input checks
+
+    wires = check_wires(wires)
+
+    check_no_variable(features, msg="'features' cannot be differentiable")
+
+    expected_shape = (len(wires),)
+    check_shape(
+        features,
+        expected_shape,
+        msg="'features' must be of shape {}; got {}" "".format(expected_shape, get_shape(features)),
+    )
+
+    check_type(
+        n_repeats, [int], msg="'n_repeats' must be an integer; got type {}".format(type(n_repeats))
+    )
+
+    check_type(
+        pattern,
+        [Iterable, type(None)],
+        msg="'pattern' must be None or a list of wire pairs; got {}".format(pattern),
+    )
+
+    if pattern is not None:
+
+        # check type
+        for p in pattern:
+            check_type(
+                p,
+                [Iterable],
+                msg="'pattern' must be None or a list of wire pairs; got {}".format(pattern),
+            )
+
+            for w in p:
+                check_type(
+                    w,
+                    [int],
+                    msg="'pattern' must be None or a list of wire pairs; got {}".format(pattern),
+                )
+
+        # check shape
+        shape = get_shape(pattern)
+        if len(shape) != 2 or shape[1] != 2:
+            raise ValueError("'pattern' must be a list of pairs of wires; got {}".format(pattern))
+
+    #####################
+
+    if pattern is None:
+        # default is an all-to-all pattern
+        pattern = [[i, j] for i in range(len(wires)) for j in range(len(wires)) if j > i]
+
+    for i in range(n_repeats):
+
+        # first block of Hadamards
+        broadcast(unitary=Hadamard, pattern="single", wires=wires)
+        # encode features into block of RZ rotations
+        broadcast(unitary=RZ, pattern="single", wires=wires, parameters=features)
+
+        # entangling block
+        products = []
+        for wire_pair in pattern:
+            # create products of parameters
+            products.append(features[wire_pair[0]] * features[wire_pair[1]])
```

Comment:
Just a comment; if the template is adapted to accept e.g. `wires = [3, 5, 7, 2]`, then this would need to be changed as well since `features[wire_pair[0]]` and `features[wire_pair[1]]` will look in the wrong places in `features`.

Reply:
You are right!

Reply:
And we cannot use the user's wire indices, because these may not exist in features!

Reply:
On the other hand, the user may _want_ the template to use a different wire order, addressing the features in that scrambled order. In which case we could just check if ``wires`` has any entry larger than its length, and leave this line of code as is. 

It is hard to tell which one is the more useful behaviour...

Reply:
I thought about it, and propose to go with the following: The i'th feature is applied to the i'th wire index (which may not be the i'th wire!).

For example: If the user passes ``features=[1., 2., 3.]`` and ``wires=[0, 17, 4]``, the first block of ``RZ``s applies:

feature 1. - > wire 0
feature 2. - > wire 17
feature 3. - > wire 4

Likewise for the entanglers!

Added a section in the docstring and a test for this.

Reply:
Sounds great; I like that solution. Thanks @mariaschuld!